File: src/relay/contract.ts
/**
 * contract.ts — Relay Protocol v1 contract manifest, deterministic SHA-256
 * contract hash, and compatibility/diff helper.
 *
 * The contract is the ChatGPT-visible MCP ABI of the workstation runtime. A
 * stable `contract_hash` lets the edge answer `initialize`/`tools/list` from a
 * frozen manifest (plan §5.3) and lets the edge/workstation refuse activation
 * when a runtime branch would silently change the model-visible ABI (§5.5,
 * §8). Runtime behavior can change freely while `contract_hash` stays equal.
 *
 * ─────────────────────────────────────────────────────────────────────────────
 * CANONICAL REPRESENTATION THE HASH COVERS (documented contract)
 * ─────────────────────────────────────────────────────────────────────────────
 * The hash is computed over the bare sorted TOOL ARRAY directly — no wrapping
 * object. The recorded epoch-1 baseline reproduces exactly:
 * `computeContractHash(baseline.tools)` === `sha256:3f23083ae31b977dad21b1ec9d6919c49e1067a27f7b7eea7bdd021b54770c0d`
 * for the 17 tools of the live `0.3.23` process in `docs/_wip/.contract-epoch1-baseline.json`.
 *
 * What the hash covers (the ChatGPT-visible ABI only):
 *
 *   tools — the tool list array, normalized as follows:
 *     - every tool sorted by `name` (ascending, byte order); tool-list order
 *       is intentionally NOT semantically meaningful, so sorting makes the
 *       hash independent of catalog ordering.
 *     - each tool contributes ALL its metadata as supplied (name, description,
 *       inputSchema, annotations, execution, etc.), NOT just a fixed subset,
 *       so contract-relevant attributes affect the hash.
 *     - duplicate tool names are rejected (a catalog with two tools sharing a
 *       name is ambiguous and cannot hash deterministically).
 *     - schema-internal array order is PRESERVED: JSON-Schema arrays such as
 *       `required`, `enum`, `oneOf` and tuple `items` are semantically
 *       order-sensitive and are never permuted.
 *
 * It explicitly DOES NOT cover (so runtime/edge bookkeeping can vary without
 * perturbing the ABI):
 *   - `contract_hash` (the value itself)
 *   - `contract_epoch`
 *   - `runtime_version`
 *   - `git_commit`
 *   - any manifest metadata outside the tool catalog (captured_at, source, …)
 *
 * Canonical encoding of the covered ABI: keys sorted recursively
 * (canonical-json.ts), tools sorted by name, arrays preserved. The result is
 * the UTF-8 bytes of that canonical JSON array fed to SHA-256, reported as
 * `sha256:<lowercase hex>`.
 * ─────────────────────────────────────────────────────────────────────────────
 */

import { createHash } from "node:crypto";
import { canonicalBytes } from "./canonical-json.js";

export const CONTRACT_HASH_PREFIX = "sha256:";
/** "sha256:" + 64 hex chars. */
export const CONTRACT_HASH_LEN = 7 + 64;

/** A tool as the workstation advertises it (model-visible ABI). */
export interface ContractTool {
  name: string;
  description?: string;
  inputSchema?: Record<string, unknown>;
  /** Additional ABI metadata (annotations, execution, etc.) — hashed too. */
  [key: string]: unknown;
}

/** Full contract manifest (model-visible ABI + identity metadata). */
export interface ContractManifest {
  contract_epoch: number;
  contract_hash: string;
  runtime_version: string;
  git_commit: string | null;
  tools: ContractTool[];
  /** Optional extra manifest fields the caller attaches; they are NOT hashed. */
  [key: string]: unknown;
}

/**
 * Normalize the tool list into the hash-covered canonical form: dedupe check
 * by name, then sort by name (byte order). Array order of the *tool list* is
 * intentionally ignored; schema-internal arrays are untouched.
 */
export function normalizeTools(tools: readonly ContractTool[]): ContractTool[] {
  const seen = new Set<string>();
  const normalized = tools.map((t) => {
    if (!t || typeof t !== "object") throw new TypeError("contract: tool must be an object");
    if (typeof t.name !== "string" || t.name.length === 0) {
      throw new TypeError("contract: each tool must have a non-empty name");
    }
    if (seen.has(t.name)) throw new Error(`contract: duplicate tool name '${t.name}'`);
    seen.add(t.name);
    return { ...t };
  });
  return normalized.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
}

/** The covered (hash-relevant) projection of a manifest: the normalized tool array itself. */
function hashProjection(tools: readonly ContractTool[]): ContractTool[] {
  return normalizeToolsForHash(tools);
}

/**
 * Build the exact canonical tool array that the hash covers. Exposed so tests and
 * tooling can inspect the deterministic input independent of hash formatting.
 */
export function contractHashSource(tools: readonly ContractTool[]): ContractTool[] {
  return hashProjection(tools);
}

function normalizeToolsForHash(tools: readonly ContractTool[]): ContractTool[] {
  return normalizeTools(tools);
}

/**
 * Compute the deterministic SHA-256 contract hash over the canonical ABI.
 * See the file header for the documented canonical representation. Output is
 * `sha256:<lowercase hex>`.
 */
export function computeContractHash(tools: readonly ContractTool[]): string {
  const source = hashProjection(tools);
  const canonical = canonicalBytes(source, 64);
  const h = createHash("sha256");
  h.update(canonical);
  return `${CONTRACT_HASH_PREFIX}${h.digest("hex")}`;
}

/**
 * Build a full manifest with a computed `contract_hash` over `tools`.
 * `runtime_version`/`git_commit`/`contract_epoch` are recorded but NOT hashed
 * (see header). Duplicate tool names throw.
 */
export function buildContractManifest(input: {
  contract_epoch: number;
  runtime_version: string;
  git_commit: string | null;
  tools: readonly ContractTool[];
}): ContractManifest {
  const tools = normalizeTools(input.tools);
  const contract_hash = computeContractHash(tools);
  return {
    contract_epoch: input.contract_epoch,
    contract_hash,
    runtime_version: input.runtime_version,
    git_commit: input.git_commit,
    tools,
  };
}

/** Verify a manifest's declared hash against its tools. */
export function verifyContractHash(manifest: ContractManifest): boolean {
  const expected = manifest.contract_hash;
  if (typeof expected !== "string" || !expected.startsWith(CONTRACT_HASH_PREFIX)) return false;
  const actual = computeContractHash(manifest.tools);
  return actual === expected;
}

/** `sha256:<hex>` syntax check (no cryptographic verification). */
export function isContractHashShape(value: unknown): value is string {
  if (typeof value !== "string") return false;
  return /^sha256:[0-9a-f]{64}$/.test(value);
}

// ─────────────────────────────────────────────────────────────────────────────
// Compatibility / diff
// ─────────────────────────────────────────────────────────────────────────────

export type ToolDiff =
  | { kind: "added"; name: string }
  | { kind: "removed"; name: string }
  | { kind: "changed"; name: string; reason: string };

export interface ContractDiff {
  equal: boolean;
  sameHash: boolean;
  sameEpoch: boolean;
  /** Added/removed/changed tools between two manifests (by name). */
  toolDiff: ToolDiff[];
}

/**
 * Compare two manifests and classify changes. The result is used to block
 * unattended runtime activation when the model-visible ABI would change.
 * `equal` is true only when epoch, hash and the tool catalog all agree.
 */
export function diffContracts(a: ContractManifest, b: ContractManifest): ContractDiff {
  const sameEpoch = a.contract_epoch === b.contract_epoch;
  const sameHash =
    typeof a.contract_hash === "string" &&
    typeof b.contract_hash === "string" &&
    a.contract_hash === b.contract_hash;
  const addedDiff: ToolDiff[] = [];
  const removedDiff: ToolDiff[] = [];
  const changedDiff: ToolDiff[] = [];
  const mapA = new Map<string, ContractTool>(a.tools.map((t) => [t.name, t]));
  const mapB = new Map<string, ContractTool>(b.tools.map((t) => [t.name, t]));
  const names = new Set([...mapA.keys(), ...mapB.keys()]);
  for (const name of names) {
    const ta = mapA.get(name);
    const tb = mapB.get(name);
    if (ta && !tb) {
      removedDiff.push({ kind: "removed", name });
    } else if (!ta && tb) {
      addedDiff.push({ kind: "added", name });
    } else if (ta && tb) {
      // Compare the hash-relevant projection (canonical tools, order-insensitive).
      const ha = computeContractHash([ta]);
      const hb = computeContractHash([tb]);
      if (ha !== hb) {
        changedDiff.push({ kind: "changed", name, reason: `canonical ABI differs (${ha} != ${hb})` });
      }
    }
  }
  const toolDiff = [...addedDiff, ...removedDiff, ...changedDiff];
  const equal = sameEpoch && sameHash && toolDiff.length === 0;
  return { equal, sameEpoch, sameHash, toolDiff };
}

File: crates/herdr-mcp/src/relay/contract.rs
//! Relay Protocol v1 contract manifest hashing (canonical SHA-256 identity).
//!
//! Faithful port of `src/relay/contract.ts`. The contract is the ChatGPT-visible
//! MCP ABI of the workstation runtime. A stable `contract_hash` lets the edge
//! answer `initialize`/`tools/list` from a frozen manifest and lets edge and
//! workstation refuse activation when a runtime branch would silently change the
//! model-visible ABI.
//!
//! Canonical representation the hash covers:
//!   - the bare sorted TOOL ARRAY directly (no wrapping object);
//!   - every tool sorted by `name` (ascending, UTF-16 code-unit / byte order);
//!     tool-list order is intentionally NOT semantically meaningful;
//!   - each tool contributes ALL metadata as supplied (name, description,
//!     inputSchema, annotations, execution, …), not a fixed subset;
//!   - duplicate tool names are rejected;
//!   - schema-internal array order is PRESERVED (`required`, `enum`, `oneOf`,
//!     tuple `items`) because those arrays are semantically order-sensitive.
//!
//! The hash explicitly does NOT cover `contract_hash`, `contract_epoch`,
//! `runtime_version`, `git_commit`, or manifest metadata outside the tool
//! catalog.

use crate::relay::canonical_json::canonical_bytes;
use serde_json::Value;
use sha2::{Digest, Sha256};

pub const CONTRACT_HASH_PREFIX: &str = "sha256:";
/// "sha256:" + 64 hex chars.
pub const CONTRACT_HASH_LEN: usize = 7 + 64;
/// Default depth cap for canonicalizing contract tool lists (matches the TS
/// `computeContractHash(tools, 64)` call site).
pub const CONTRACT_MAX_DEPTH: usize = 64;

/// `sha256:<hex>` syntax check (no cryptographic verification).
pub fn is_contract_hash_shape(value: &Value) -> bool {
    let Some(s) = value.as_str() else {
        return false;
    };
    is_contract_hash_shape_str(s)
}

/// `sha256:<hex>` syntax check over a plain string.
pub fn is_contract_hash_shape_str(s: &str) -> bool {
    s.len() == CONTRACT_HASH_LEN
        && s.starts_with(CONTRACT_HASH_PREFIX)
        && s[7..]
            .bytes()
            .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
}

/// Normalize the tool list into the hash-covered canonical form: dedupe check
/// by name, then sort by name. Schema-internal arrays are untouched.
pub fn normalize_tools(tools: &[Value]) -> Result<Vec<Value>, String> {
    let mut seen = std::collections::HashSet::new();
    let mut normalized = Vec::with_capacity(tools.len());
    for t in tools {
        if !t.is_object() {
            return Err("contract: tool must be an object".to_owned());
        }
        let name = t
            .get("name")
            .and_then(Value::as_str)
            .ok_or_else(|| "contract: each tool must have a non-empty name".to_owned())?;
        if name.is_empty() {
            return Err("contract: each tool must have a non-empty name".to_owned());
        }
        if !seen.insert(name) {
            return Err(format!("contract: duplicate tool name '{name}'"));
        }
        normalized.push(t.clone());
    }
    normalized.sort_by(|a, b| {
        let an = a.get("name").and_then(Value::as_str).unwrap_or_default();
        let bn = b.get("name").and_then(Value::as_str).unwrap_or_default();
        an.encode_utf16().cmp(bn.encode_utf16())
    });
    Ok(normalized)
}

/// Compute the deterministic SHA-256 contract hash over the canonical ABI.
/// Output is `sha256:<lowercase hex>`.
pub fn compute_contract_hash(tools: &[Value]) -> Result<String, String> {
    let source = normalize_tools(tools)?;
    let bytes = canonical_bytes(&Value::Array(source), CONTRACT_MAX_DEPTH)
        .map_err(|e| format!("contract: cannot canonicalize tools: {e}"))?;
    let digest = Sha256::digest(&bytes);
    Ok(format!("{CONTRACT_HASH_PREFIX}{digest:x}"))
}

/// Verify a declared `sha256:<hex>` hash string against a tool list.
pub fn verify_contract_hash(declared_hash: &str, tools: &[Value]) -> bool {
    if !is_contract_hash_shape_str(declared_hash) {
        return false;
    }
    match compute_contract_hash(tools) {
        Ok(actual) => actual == declared_hash,
        Err(_) => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn is_contract_hash_shape_accepts_hex_and_rejects_others() {
        assert!(is_contract_hash_shape_str(&format!(
            "sha256:{}",
            "ab".repeat(32)
        )));
        assert!(!is_contract_hash_shape_str(&format!(
            "sha256:{}",
            "ab".repeat(31)
        )));
        assert!(!is_contract_hash_shape_str(&format!(
            "sha256:{}",
            "AB".repeat(32)
        )));
        assert!(!is_contract_hash_shape_str("sha256:nothex"));
        assert!(!is_contract_hash_shape(&Value::Null));
        assert!(!is_contract_hash_shape(&json!(42)));
    }

    #[test]
    fn duplicate_tool_names_are_rejected() {
        let tools = json!([
            {"name": "dup", "description": "first"},
            {"name": "dup", "description": "second"}
        ]);
        let arr = tools.as_array().unwrap();
        assert!(compute_contract_hash(arr).is_err());
        assert!(normalize_tools(arr).is_err());
        assert!(!verify_contract_hash(
            "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            arr
        ));
    }

    #[test]
    fn hash_independent_of_tool_list_ordering_but_sensitive_to_schema_array_order() {
        let a = json!([
            {"name": "herdr_a", "description": "A", "inputSchema": {"required": ["x"], "properties": {"x": {"type": "string"}}}},
            {"name": "herdr_b", "description": "B", "inputSchema": {"properties": {"y": {"type": "number"}}}}
        ]);
        let a_arr = a.as_array().unwrap();
        let mut b = a_arr.clone();
        b.reverse();
        assert_eq!(
            compute_contract_hash(a_arr).unwrap(),
            compute_contract_hash(&b).unwrap()
        );

        let ordered = json!([{"name": "herdr_a", "inputSchema": {"required": ["a", "b"], "properties": {"a": {}, "b": {}}}}]);
        let reversed = json!([{"name": "herdr_a", "inputSchema": {"required": ["b", "a"], "properties": {"b": {}, "a": {}}}}]);
        assert_ne!(
            compute_contract_hash(ordered.as_array().unwrap()).unwrap(),
            compute_contract_hash(reversed.as_array().unwrap()).unwrap()
        );
    }

    #[test]
    fn hash_independent_of_object_key_insertion_order_within_tool() {
        let t1 = json!({"name": "t", "description": "d", "inputSchema": {"type": "object", "properties": {"a": {"type": "string"}}}});
        let t2 = json!({"inputSchema": {"properties": {"a": {"type": "string"}}, "type": "object"}, "description": "d", "name": "t"});
        assert_eq!(
            compute_contract_hash(&[t1]).unwrap(),
            compute_contract_hash(&[t2]).unwrap()
        );
    }

    #[test]
    fn annotations_change_the_hash() {
        let base = json!({"name": "t", "description": "d", "inputSchema": {"type": "object"}});
        let annotated = json!({"name": "t", "description": "d", "inputSchema": {"type": "object"}, "annotations": {"readOnlyHint": true}});
        assert_ne!(
            compute_contract_hash(&[base]).unwrap(),
            compute_contract_hash(&[annotated]).unwrap()
        );
    }

    #[test]
    fn epoch2_public_contract_hash_is_exact() {
        // Embedded frozen public catalog; the canonical hash must be byte-exact.
        let catalog = crate::contract::tool_catalog().expect("embedded epoch2 tools");
        let expected = "sha256:7da23ad2ec8e7703d6380062126ba797218bde9e7711138c6b3e0ca6592efbf8";
        assert_eq!(compute_contract_hash(&catalog).unwrap(), expected);
        assert!(verify_contract_hash(expected, &catalog));
        // Order independence: a reversed catalog yields the same hash.
        let mut rev = catalog.clone();
        rev.reverse();
        assert_eq!(compute_contract_hash(&rev).unwrap(), expected);
    }
}
